stdlib/public/TensorFlow/PythonConversion.swift

```diff
@@ -164,4 +164,21 @@ extension Tensor where Scalar : NumpyScalarCompatible {
   public func makeNumpyArray() -> PythonObject { return array.makeNumpyArray() }
 }
 
+extension TensorShape : PythonConvertible {
+  public var pythonObject: PythonObject {
+    return dimensions.pythonObject
+  }
+
+  public init?(_ pythonObject: PythonObject) {
```

Comment:
I didn't review the last change closely. Maybe tests broke because `TensorShape(0)`'s initialization defaulted to `TensorShape(0 as PythonObject)` somehow?

Reply:
@rxwei Yep, that is exactly the case.  I spent some time looking into how to manipulate the  prority order of the initializers, but if you have a suggested approach i can definitely try it out.  I settled for just making sure that this initializer successfully passes through to the correct initializer, since ```TensorShape(0 as PythonObject)``` should be valid anyway.

It is also possible that the failure was due to this initializer being failable, but the only reliable repro steps are what I outlined [here](https://bugs.swift.org/browse/TF-426).

I guess the question is what would you want/expect the output of the below program to be?
```
func doThing(_ arg: PythonObject) {
  print(type(of: arg))
}

func doThing(_ arg: Int32) {
  print(type(of: arg))
}

doThing(1)
```

Output currently is PythonObject.  Perhaps that is the real underlying bug.

Reply:
I see. There are two bigger problems. First, the TensorFlow module shouldn't logically depend on Python, i.e. there can be targets that require all APIs that use Python be `#if`'d out. Second, we ultimately want to optimize tensor operations through techniques like graph program extraction, but having this Python layer for the initialization of a simple tensor shape makes it extremely inefficient when we are targeting hardware like TPUs.

I think a better solution would be to split this protocol into two. This has actually been on my mind for over a year now:

`PythonConvertible` defines conversion **to** Python; `ConvertibleFromPython` defines conversion **from** Python.

`TensorShape` will then only conform to `PythonConvertible`, but not the other one, making this integer literal issue go away.

Reply:
Ah, makes sense.  I could definitely work on a PR for the split up of the protocol.

Reply:
Great. Let's do that!